drivers/serial/uart_ns16550.c

```diff
@@ -461,12 +453,33 @@ static int uart_ns16550_config_get(struct device *dev, struct uart_config *cfg)
  */
 static int uart_ns16550_init(struct device *dev)
 {
+	struct uart_ns16550_dev_data_t * const dev_data = DEV_DATA(dev);
+	const struct uart_ns16550_device_config * const dev_cfg = DEV_CFG(dev);
+
+	ARG_UNUSED(dev_data);
+	ARG_UNUSED(dev_cfg);
+
+#ifdef UART_NS16550_PCIE_ENABLED
+	if (dev_cfg->pcie) {
+		if (!pcie_probe(dev_cfg->pcie_bdf, dev_cfg->pcie_id)) {
+			return -EINVAL;
+		}
+
+		dev_data->port = pcie_get_mbar(dev_cfg->pcie_bdf, 0);
+		pcie_set_cmd(dev_cfg->pcie_bdf, PCIE_CONF_CMDSTAT_MEM, true);
+	}
+#endif
+
 	uart_ns16550_configure(dev, &DEV_DATA(dev)->uart_config);
 
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 	DEV_CFG(dev)->devconf.irq_config_func(dev);
 #endif
 
+#ifndef UART_NS16550_PCIE_ENABLED
+	DEV_DATA(dev)->port = DEV_CFG(dev)->devconf.port;
```

Comment:
This is not right... If PCIE is not enabled, `DEV_DATA(dev)->port` is not set when calling `uart_ns16550_configure()` above. So none of the configuration is being done. And that would be accessing I/O port @ 0x0000, or memory @ 0x00000000 + offset.

Reply:
@dcpleung Thanks for looking at it. That is interesting. Perhaps that is because of the recent change implemented as @pabigot suggested as a resolution to the handling of port setting:
> If you fix that (i.e. switch all the uses back to DEV_DATA) then you'll need to set DEV_DATA(dev)->port from DEV_CFG(dev)->devconf.port in the init function.

Reply:
I'll need to look at this more